### PR TITLE
add base and excel connector

### PR DESCRIPTION
**Pull request by refarer**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,52 +1,16 @@
...
-5. `npm run build:watch`
-6. Implement your events/actions in `src/index.ts`
-7. `npm run lint`
-8. Push your code
-9. Go to https://app.circleci.com/projects/project-dashboard/github/reshufflehq/
-    a. You should see your new connector repo
-    b. click on `Set Up Project` for the repo
-    c. click on `Use Existing Config`
-    d. click on `Start Building`
+This package contains [Reshuffle](https://github.com/reshufflehq/reshuffle) connectors to (some of the) services available from Microsoft. These connectors are all implemented using the [Microsoft Graph SDK for JS](https://github.com/microsoftgraph/msgraph-sdk-javascript).
```
Comment (foundingnimo) [inline README.md]: I think we should remove the (..) around `some of the` or take the `the` out of them

File: doc/ExcelConnector.md [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+excelConnector
+  .addNewWorksheet('drive/items/{item-id}', 'myNewSheet')
+  .then((newWorksheet) => console.log(newWorksheet))
+```
+
+#### Table of Contents
+
+[Configuration](#ConfigurationOptions) Configuration options
+
+[Types](#TypeScriptTypes) TypeScript types
```
Comment (foundingnimo) [inline doc/ExcelConnector.md]: These read a bit weird IMO. I think you can take the text on the far right (eg `Configuration options`) and use it as the link text - so `[Configuration options](#ConfigurationOptions)` etc

File: doc/ExcelConnector.md [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+async getTables(driveItem: string, name: string): Promise<MicrosoftGraph.WorkbookTable[]>
+```
+
+##### getTables
+
+Get Tables.
+
+```ts
+async getTables(driveItem: string, name: string): Promise<MicrosoftGraph.WorkbookTable[]>
+```
```
Comment (foundingnimo) [inline doc/ExcelConnector.md]: Duplicate?

File: doc/ExcelConnector.md [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+
+##### insertRange
+
+Insert a Range.
+
+```ts
+async insertRange(
+    driveItem: string,
+    name: string,
```
Comment (foundingnimo) [inline doc/ExcelConnector.md:186]: Does the insert not have an address?

File: doc/ExcelConnector.md [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+    values?: Record<string, unknown>,
+    formula?: Record<string, unknown>,
+    numberFormat?: Record<string, unknown>,
+  ): Promise<MicrosoftGraph.WorkbookRange>
+```
```
Comment (foundingnimo) [inline doc/ExcelConnector.md]: Duplicate?

File: package.json [hunk 6]
```diff
@@ -13,7 +13,7 @@
   },
   "repository": {
     "type": "git",
-    "url": "git+https://github.com/reshufflehq/reshuffle-_CONNECTOR_NAME_-connector.git"
+    "url": "git+https://github.com/reshufflehq/reshuffle-microsoft-connector.git"
```
Comment (foundingnimo) [inline package.json]: should this be `git+https://github.com/reshufflehq/reshuffle-microsoft-connector`**s**`.git` ?

(Also 2 more instances below)

File: src/auth.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+            client_secret: this.options.AppPassword,
+            scope: 'https://graph.microsoft.com/.default',
+            grant_type: 'client_credentials',
+          }),
+        },
+      )
+      if (response.ok) {
+        return (await response.json()).access_token
+      } else {
+        throw new Error()
```
Comment (foundingnimo) [inline src/auth.ts]: We should probably at least log the response if it's not ok, and possibly add info to the error?

File: src/base.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+  MicrosoftConnectorConfigOptions,
+  undefined
+> {
+  client: Client
+
+  constructor(app: Reshuffle, options: MicrosoftConnectorConfigOptions, id?: string) {
+    super(app, options, id)
+    const clientOptions: ClientOptions = {
+      defaultVersion: 'v1.0',
+      debugLogging: false,
```
Comment (foundingnimo) [inline src/base.ts]: This should probably be an env variable with default=false

File: src/excel.ts [hunk 9]
```diff
@@ -0,0 +1,104 @@
+import './fetch'
+import MicrosoftGraph from '@microsoft/microsoft-graph-types'
+
+import { MicrosoftConnector } from './base'
+
+export default class ExcelConnector extends MicrosoftConnector {
+  async getDriveItem(driveItem: string): Promise<MicrosoftGraph.DriveItem> {
```
Comment (foundingnimo) [inline src/excel.ts:7]: This and other actions - we should think about try...catching the requests

File: src/excel.ts [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    name: string,
+    newName?: string,
+    position?: number,
+    visibility?: 'Visible' | 'Hidden' | 'VeryHidden',
+  ): Promise<MicrosoftGraph.WorkbookWorksheet> {
+    const request = await this.client
+      .api(`${driveItem}/workbook/worksheets/${name}`)
+      .update({ position, name: newName, visibility })
+    return request
+  }
```
Comment (foundingnimo) [inline src/excel.ts:51]: Why the change in formatting from `updateWorksheet` onwards?

File: src/excel.ts [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+    values?: Record<string, unknown>,
+    formula?: Record<string, unknown>,
+    numberFormat?: Record<string, unknown>,
+  ): Promise<MicrosoftGraph.WorkbookRange> {
+    const request = await this.client
+      .api(`${driveItem}/workbook/worksheets/${name}/range/insert`)
+      .update({ values, formula, numberFormat })
+    return request
+  }
+}
```
Comment (foundingnimo) [inline src/excel.ts]: This is a copy of `updateRange` - but the API for insert seems different: https://docs.microsoft.com/en-us/graph/api/range-insert?view=graph-rest-1.0&tabs=http - need a shift value
(Though I don't understand from the documentation how we identify where the range is inserted?)

File: src/fetch.ts [hunk 12]
```diff
@@ -0,0 +1,13 @@
+import _fetch from 'node-fetch'
+
+declare global {
+  const fetch: typeof _fetch
+}
+
+// @ts-ignore
+if (!globalThis.fetch) {
+  // @ts-ignore
+  globalThis.fetch = _fetch
```
Comment (foundingnimo) [inline src/fetch.ts:10]: Do we need this for a node application?

File: src/index.ts [hunk 13]
```diff
@@ -1,61 +1,4 @@
...
-    // Your implementation here
-  }
-
-  action2(foo: string): void {
-    // Your implementation here
-  }
-}
-
-export { _CONNECTOR_NAME_Connector }
+export { MicrosoftConnector, ExcelConnector }
```
Comment (foundingnimo) [inline src/index.ts]: I think we can just go
```
export * from './excel'
export * from './base'
```

File: src/base.ts [hunk 14]
```diff
@@ -0,0 +1,29 @@
+import { Reshuffle, BaseHttpConnector } from 'reshuffle-base-connector'
+import { Client, ClientOptions } from '@microsoft/microsoft-graph-client'
+
+import { ClientCredentialAuthenticationProvider, MicrosoftOptions } from './auth'
+
+export type MicrosoftConnectorConfigOptions = MicrosoftOptions
```
Comment (foundingnimo) [inline src/base.ts]: Not used outside this here and not exported in `/src/index.ts` as well?

File: src/base.ts [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+    }
+    this.client = Client.initWithMiddleware(clientOptions)
+  }
+
+  sdk(): Client {
+    return this.client
+  }
+}
+
+export { MicrosoftConnector }
```
Comment (foundingnimo) [inline src/base.ts]: Is this required? (We have `export default` above?)

**Review by foundingnimo — CHANGES_REQUESTED** [review]

File: src/fetch.ts (hunk 12 above)
Comment (refarer) [reply src/fetch.ts:10]: Node doesn't have fetch and `msgraph-sdk-javascript` needs fetch to be poly-filled, I could just use the `isomorphic-fetch` package, but we don't need the poly-fill for web also, so I just did it this way.

File: src/base.ts (hunk 14 above)
Comment (refarer) [reply src/base.ts]: Will export it from `index.ts`

File: src/excel.ts (hunk 10 above)
Comment (refarer) [reply src/excel.ts:51]: It's just the Prettier config

File: package.json (hunk 6 above)
Comment (foundingnimo) [reply package.json]: ((Also 2 more instances below))

- Also - the description is 'a Microsoft connector'

**Review by foundingnimo — APPROVED** [review]
Approved pending the last 3 changes in `package.json`

**Review by foundingnimo — APPROVED** [review]
